File: scuttle_mapping/scuttle_mapping/mapping_with_known_poses.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from sensor_msgs.msg import LaserScan
from nav_msgs.msg import OccupancyGrid, MapMetaData
from tf2_ros import Buffer, TransformListener
from tf2_ros import LookupException, ConnectivityException, ExtrapolationException
from tf_transformations import euler_from_quaternion
import math
import numpy as np

# Log-Odds Constants
# A higher probability > 0.5 means occupied, < 0.5 means free
PRIOR_PROB = 0.5
OCC_PROB = 0.9
FREE_PROB = 0.35

class Pose:
    def __init__(self, px=0, py=0):
        self.x = px
        self.y = py

def prob2logodds(p):
    return math.log(p / (1 - p))

def logodds2prob(l):
    try:
        return 1 - (1 / (1 + math.exp(l)))
    except OverflowError:
        return 1.0 if l > 0 else 0.0

def poseToCell(pose: Pose, map_info: MapMetaData):
    return map_info.width * pose.y + pose.x

def coordinatesToPose(px, py, map_info: MapMetaData):
    pose = Pose()
    pose.x = round((px - map_info.origin.position.x) / map_info.resolution)
    pose.y = round((py - map_info.origin.position.y) / map_info.resolution)
    return pose

def poseOnMap(pose: Pose, map_info: MapMetaData):
    return pose.x < map_info.width and pose.x >= 0 and pose.y < map_info.height and pose.y >= 0

def bresenham(start: Pose, end: Pose):
    """
    Standard Bresenham's Line Algorithm to find all grid cells 
    intersected by the laser ray.
    """
    line = []

    dx = end.x - start.x
    dy = end.y - start.y

    xsign = 1 if dx > 0 else -1
    ysign = 1 if dy > 0 else -1

    dx = abs(dx)
    dy = abs(dy)

    if dx > dy:
        xx = xsign
        xy = 0
        yx = 0
        yy = ysign
    else:
        tmp = dx
        dx = dy
        dy = tmp
        xx = 0
        xy = ysign
        yx = xsign
        yy = 0

    D = 2 * dy - dx
    y = 0

    for i in range(dx + 1):
        line.append(Pose(start.x + i * xx + y * yx, start.y + i * xy + y * yy))
        if D >= 0:
            y += 1
            D -= 2 * dx
        D += 2 * dy

    return line

def inverseSensorModel(p_robot: Pose, p_beam: Pose):
    """
    Returns a list of (Pose, Probability) tuples.
    Cells along the ray are FREE. The endpoint is OCCUPIED.
    """
    occ_values = []
    line = bresenham(p_robot, p_beam)

    # All cells along the ray (excluding the last one) are likely free
    for pose in line[:-1]:
        occ_values.append((pose, FREE_PROB))

    # The final cell where the laser hit is likely occupied
    occ_values.append((line[-1], OCC_PROB))
    return occ_values

class MappingWithKnownPoses(Node):
    def __init__(self):
        super().__init__("mapping_with_known_poses")
        
        # Parameters for Map Size
        self.declare_parameter("width", 50.0)   # Meters
        self.declare_parameter("height", 50.0)  # Meters
        self.declare_parameter("resolution", 0.1) # Meters per pixel

        width = self.get_parameter("width").value
        height = self.get_parameter("height").value
        resolution = self.get_parameter("resolution").value

        # Initialize Occupancy Grid
        self.map_ = OccupancyGrid()
        self.map_.info.resolution = resolution
        self.map_.info.width = round(width / resolution)
        self.map_.info.height = round(height / resolution)
        # Center the map so (0,0) is in the middle
        self.map_.info.origin.position.x = float(-round(width / 2.0))
        self.map_.info.origin.position.y = float(-round(height / 2.0))
        self.map_.header.frame_id = "odom"
        
        # Initialize grid data to -1 (Unknown)
        self.map_.data = [-1] * (self.map_.info.width * self.map_.info.height)
        
        # Internal probability representation (using log odds)
        self.probability_map_ = [prob2logodds(PRIOR_PROB)] * (self.map_.info.width * self.map_.info.height)

        # TF and Subs/Pubs
        self.tf_buffer = Buffer()
        self.tf_listener = TransformListener(self.tf_buffer, self)
        
        self.scan_sub = self.create_subscription(LaserScan, "scan", self.scanCallback, 10)
        self.map_pub = self.create_publisher(OccupancyGrid, "map", 1)
        
        # Timer to publish the map periodically (every 2 seconds to save CPU)
        self.timer = self.create_timer(2.0, self.timerCallback)
        
        self.get_logger().info("Mapping Node Started. Waiting for Scan data...")

    def scanCallback(self, scan: LaserScan):
        try:
            # Look up the transform from map frame (odom) to sensor frame
            # We use Time() to get the latest available transform
            t = self.tf_buffer.lookup_transform(
                self.map_.header.frame_id, 
                scan.header.frame_id, 
                rclpy.time.Time())
        except (LookupException, ConnectivityException, ExtrapolationException):
            self.get_logger().warn("Could not transform scan to odom frame.")
            return

        # Robot Pose in Grid Coordinates
        robot_p = coordinatesToPose(t.transform.translation.x, t.transform.translation.y, self.map_.info)
        
        if not poseOnMap(robot_p, self.map_.info):
            self.get_logger().warn("Robot is outside the map boundaries!")
            return

        # Robot Yaw
        (roll, pitch, yaw) = euler_from_quaternion(
            [t.transform.rotation.x, t.transform.rotation.y,
             t.transform.rotation.z, t.transform.rotation.w])
        
        # Process every Nth ray to save CPU (Python is slow)
        step = 5 
        for i in range(0, len(scan.ranges), step):
            r = scan.ranges[i]
            
            # Ignore invalid ranges
            if math.isinf(r) or math.isnan(r) or r > scan.range_max or r < scan.range_min:
                continue

            # Polar to Cartesian conversion (Global Frame)
            # angle = sensor_angle + robot_yaw
            angle = scan.angle_min + (i * scan.angle_increment) + yaw
            
            px = r * math.cos(angle) + t.transform.translation.x
            py = r * math.sin(angle) + t.transform.translation.y

            # Beam Endpoint in Grid Coordinates
            beam_p = coordinatesToPose(px, py, self.map_.info)
            
            if not poseOnMap(beam_p, self.map_.info):
                continue

            # Raytrace
            poses = inverseSensorModel(robot_p, beam_p)

            # Update Log Odds
            for pose, value in poses:
                if poseOnMap(pose, self.map_.info):
                    cell = poseToCell(pose, self.map_.info)
                    self.probability_map_[cell] += prob2logodds(value) - prob2logodds(PRIOR_PROB)

    def timerCallback(self):
        # Convert Log Odds back to 0-100 probability for visualization
        self.map_.data = [int(logodds2prob(value) * 100) for value in self.probability_map_]
        self.map_.header.stamp = self.get_clock().now().to_msg()
        self.map_pub.publish(self.map_)

def main(args=None):
    rclpy.init(args=args)
    node = MappingWithKnownPoses()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: scuttle_mapping/src/mapping_with_known_poses.cpp
#include "scuttle_mapping/mapping_with_known_poses.hpp"

#include <chrono>
#include <algorithm>
#include <cmath>

#include "geometry_msgs/msg/transform_stamped.hpp"
#include "tf2/utils.h"
#include "tf2/LinearMath/Quaternion.h"
#include "tf2/LinearMath/Matrix3x3.h"

// Using namespaces
using namespace std;
using namespace std::chrono_literals;
using namespace scuttle_mapping;
using namespace rclcpp;
using namespace sensor_msgs::msg;
using namespace nav_msgs::msg;
using namespace geometry_msgs::msg;

namespace scuttle_mapping
{

double prob2logodds(double p)
{
    return log(p / (1 - p));
}

double logodds2prob(double l)
{
    return 1 - (1 / (1 + exp(l)));
}

unsigned int poseToCell(const Pose & pose, const MapMetaData & map_info)
{
    return map_info.width * pose.y + pose.x;
}

Pose coordinatesToPose(const double px, const double py, const MapMetaData & map_info)
{
    Pose pose;
    pose.x = round((px - map_info.origin.position.x) / map_info.resolution);
    pose.y = round((py - map_info.origin.position.y) / map_info.resolution);
    return pose;
}

bool poseOnMap(const Pose & pose, const MapMetaData & map_info)
{
    return pose.x < static_cast<int>(map_info.width)  && pose.x >= 0 &&
           pose.y < static_cast<int>(map_info.height) && pose.y >= 0;
}

vector<Pose> bresenham(const Pose & start, const Pose & end)
{
    vector<Pose> line;

    int dx = end.x - start.x;
    int dy = end.y - start.y;

    int xsign = dx > 0 ? 1 : -1;
    int ysign = dy > 0 ? 1 : -1;

    dx = abs(dx);
    dy = abs(dy);

    int xx, xy, yx, yy;
    if(dx > dy)
    {
        xx = xsign; xy = 0;
        yx = 0;     yy = ysign;
    }
    else
    {
        int tmp = dx;
        dx = dy;
        dy = tmp;
        xx = 0;     xy = ysign;
        yx = xsign; yy = 0;
    }

    int D = 2 * dy - dx;
    int y = 0;

    line.reserve(dx + 1);
    for (int i = 0; i < dx + 1; i++)
    {
        line.emplace_back(Pose(start.x + i * xx + y * yx, start.y + i * xy + y * yy));
        if(D >= 0)
        {
            y++;
            D -= 2 * dx;
        }
        D += 2 * dy;
    }

    return line;
}

vector<pair<Pose, double>> inverseSensorModel(const Pose & p_robot, const Pose & p_beam)
{
    vector<pair<Pose, double>> occ_values;
    vector<Pose> line = bresenham(p_robot, p_beam);
    
    if (line.empty()) return occ_values;

    occ_values.reserve(line.size());

    // Free space along the ray
    for (size_t i = 0; i < line.size() - 1u; i++)
    {
        occ_values.emplace_back(make_pair(line.at(i), FREE_PROB)); 
    }

    // Occupied at the end
    occ_values.emplace_back(make_pair(line.back(), OCC_PROB));
    return occ_values;
}

MappingWithKnownPoses::MappingWithKnownPoses(const string &name)
    : Node(name)
{
    declare_parameter<double>("width", 50.0);
    declare_parameter<double>("height", 50.0);
    declare_parameter<double>("resolution", 0.1);

    double width = get_parameter("width").as_double();
    double height = get_parameter("height").as_double();
    map_.info.resolution = get_parameter("resolution").as_double();
    
    map_.info.width = round(width / map_.info.resolution);
    map_.info.height = round(height / map_.info.resolution);
    map_.info.origin.position.x = - round(width / 2.0);
    map_.info.origin.position.y = - round(height / 2.0);
    map_.header.frame_id = "odom";

    // Init map with prior probability (log odds)
    map_.data = vector<int8_t>(map_.info.height * map_.info.width, -1);
    probability_map_ = vector<double>(map_.info.height * map_.info.width, prob2logodds(PRIOR_PROB));

    // FIX: Use std::make_unique to avoid conflict with rclcpp::Node::make_unique
    tf_buffer_ = std::make_unique<tf2_ros::Buffer>(this->get_clock());
    // FIX: Use std::make_shared to avoid conflict with rclcpp::Node::make_shared
    tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);
    
    scan_sub_ = create_subscription<LaserScan>(
        "scan", 10, bind(&MappingWithKnownPoses::scanCallback, this, std::placeholders::_1));
        
    map_pub_ = create_publisher<OccupancyGrid>("map", 1);
    
    timer_ = create_wall_timer(2s, bind(&MappingWithKnownPoses::timerCallback, this));
    
    RCLCPP_INFO(get_logger(), "Mapping Node Started (C++).");
}

void MappingWithKnownPoses::scanCallback(const LaserScan &scan)
{
    TransformStamped t;
    try
    {
        // Lookup transform from Map Frame (odom) to Sensor Frame
        t = tf_buffer_->lookupTransform(
            map_.header.frame_id, 
            scan.header.frame_id, 
            tf2::TimePointZero);
    }
    catch (const tf2::TransformException &ex)
    {
        RCLCPP_WARN_THROTTLE(get_logger(), *get_clock(), 1000, 
            "Could not transform %s to %s: %s", 
            scan.header.frame_id.c_str(), 
            map_.header.frame_id.c_str(), 
            ex.what());
        return;
    }

    // Check if robot pose is on the map
    Pose robot_p = coordinatesToPose(t.transform.translation.x, t.transform.translation.y, map_.info);
    if(!poseOnMap(robot_p, map_.info))
    {
        RCLCPP_WARN_THROTTLE(get_logger(), *get_clock(), 2000, "Robot is out of map bounds!");
        return;
    }

    // Get Yaw
    tf2::Quaternion q(t.transform.rotation.x, t.transform.rotation.y, t.transform.rotation.z, t.transform.rotation.w);
    tf2::Matrix3x3 m(q);
    double roll, pitch, yaw;
    m.getRPY(roll, pitch, yaw);

    // Raytracing
    // Optional: Add 'step' here (e.g. i+=5) if CPU usage is too high
    for (size_t i = 0; i < scan.ranges.size(); i++)
    {
      float r = scan.ranges.at(i);
      
      // Filter invalid ranges
      if (std::isinf(r) || std::isnan(r) || r < scan.range_min || r > scan.range_max)
          continue;

      // Polar to cartesian coordinates
      double angle = scan.angle_min + (i * scan.angle_increment) + yaw;
      double px = r * cos(angle) + t.transform.translation.x;
      double py = r * sin(angle) + t.transform.translation.y;

      Pose beam_p = coordinatesToPose(px, py, map_.info);
      
      // Skip if beam lands outside map
      if(!poseOnMap(beam_p, map_.info))
      {
        continue;
      }
      
      vector<pair<Pose, double>> poses = inverseSensorModel(robot_p, beam_p);

      for(const auto & pose : poses)
      {
        if(poseOnMap(pose.first, map_.info))
        {
            unsigned int cell = poseToCell(pose.first, map_.info);
            // Update Log Odds
            probability_map_.at(cell) += prob2logodds(pose.second) - prob2logodds(PRIOR_PROB);
        }
      }
    }
}

void MappingWithKnownPoses::timerCallback()
{
    // Convert Log Odds to occupancy grid (0-100) and publish
    // This runs efficiently every 2 seconds
    for (size_t i = 0; i < probability_map_.size(); ++i) {
        map_.data[i] = static_cast<int8_t>(logodds2prob(probability_map_[i]) * 100);
    }
    
    map_.header.stamp = get_clock()->now();
    map_pub_->publish(map_);
}

}  // namespace scuttle_mapping

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    auto node = make_shared<scuttle_mapping::MappingWithKnownPoses>("mapping_with_known_poses");
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}